colcon_core/argument_parser/type_collector.py

```diff
@@ -0,0 +1,91 @@
+# Copyright 2020 Scott K Logan
+# Licensed under the Apache License, Version 2.0
+
+from colcon_core.argument_parser import ArgumentParserDecorator
+
+
+class TypeCollectorDecorator(ArgumentParserDecorator):
+    """Collect the type conversions of arguments."""
+
+    def __init__(self, parser, **kwargs):
+        """
+        Construct a TypeCollectorDecorator.
+
+        :param parser: The argument parser to decorate
+        """
+        # avoid setting members directly, the base class overrides __setattr__
+        # pass them as keyword arguments instead
+        super().__init__(
+            parser,
+            _registered_types={},
+            **kwargs)
+
+    def get_types(self):
+        """
+        Get types for all added arguments.
+
+        :returns: The registered type converters
+        :rtype: dict
+        """
+        return dict(self._registered_types)
+
+    def add_argument(self, *args, **kwargs):
+        """Collect type converters for all added arguments."""
+        ret = super().add_argument(*args, **kwargs)
+
+        type_arg = kwargs.get('type')
+        if type_arg is not None:
+            self._registered_types[type_arg] = type_arg
```

Comment:
As I was reviewing I questioned why saving the type_arg as both the key and the value this way was necessary. As far as I understand after completing the review.

This will register "builtin" argparse types such as `int`, `float`, etc where the type and converting function are the same.

When registering new/custom types explicitly via the decorated register call, the type value name and converting function (object?) are used as the key and value respectively.

Assuming that this is indeed the case, is there any protection against add_argument "overwriting" a type registration?


Looking at this example from the tests: https://github.com/colcon/colcon-core/pull/416/files#diff-542dd1edd503344fa7bc738dc8826d89422810804591f59c35fa60b34daa25c7R28-R34

Why doesn't `p.add_argument('-x', type='hex'...)` override the type registration stored during `p.register('type', 'hex',...)`?

Reply:
This is not very discoverable, and it took me some time to really understand. The `type` for `add_argument` is first regarded as a key into the mapping of registered types. The data type of the key doesn't matter at all, though strings are the most common. The fallback for when the key wasn't found in the registry is the `type` itself, assuming it is callable. If it isn't, you'll get an exception.

This line is taking advantage of the fact that no regard is given to the data type of the key, meaning a callable can be used to index into the registry. Whenever a conversion is attempted for, say, `int()`, I've mapped that function to a no-op during suppression.

The code here results in the parser's state differing slightly after suppression is reverted. Since there's no API for removing stuff from the type registry (or at least there wasn't when I wrote this), I'm just adding registry entries that map the callables passed to `add_argument` back to themselves. This should result in the same behavior as if they were missing entirely (and the default was returned).

This is the relevant code in argparse: https://github.com/python/cpython/blob/b2ec37a7224318d466e0877790ed740ce8b4867d/Lib/argparse.py#L1418-L1421

Reply:
I re-read your question, and I understand what you're getting at now.

This is a bug! Fix (and test) incoming.

Reply:
Should be fixed by 57a299a.